dataplane/src/drivers/kernel.rs

```diff
@@ -130,100 +137,217 @@ fn get_interface_ifindex(interfaces: &[Interface], name: &str) -> Option<u32> {
+                }
             } else {
                 warn!("Could not find ifindex of interface '{name}'");
             }
         }
     }
 
-    kiftable
+    Ok(kiftable)
 }
 
 /// Main structure representing the kernel driver.
-/// This version of the kernel driver does not create any interface,
-/// but expects to be indicated the interfaces to receive packets from.
+/// This driver:
+///  * receives raw frames via `AF_PACKET``, parses to `Packet<TestBuffer>`
+///  * selects a worker by symmetric flow hash
+///  * workers run independent pipelines and send processed packets back
+///  * dispatcher serializes & transmits on the chosen outgoing interface
 pub struct DriverKernel;
+
 impl DriverKernel {
-    /// Starts the kernel driver
-    // After doing the thread spawning, this should allow should not be needed anymore
-    #[allow(clippy::needless_pass_by_value)]
+    /// Compute a **symmetric** worker index for a parsed `Packet` using a bidirectional flow key.
+    #[must_use]
+    pub fn compute_worker_idx(pkt: &Packet<TestBuffer>, workers: usize) -> usize {
+        let n = workers.max(1);
+        // Try flow-key-based symmetric hash (A<->B map to same bucket)
+        if let Ok(flow_key) = FlowKey::try_from(Bidi(pkt)) {
+            let mut h = AHasher::default();
+            flow_key.hash(&mut h);
+            let hv = usize::try_from(h.finish());
+            if n.is_power_of_two() {
+                hv & (n - 1)
+            } else {
+                hv % n
+            };
+        }
+        // TODO: Fallback for vlan
+        0
+    }
+
+    /// Spawn `workers` processing threads, each with its own pipeline instance.
+    ///
+    /// Returns:
+    ///   - `Vec<Sender<Packet<TestBuffer>>>` one sender per worker (dispatcher -> worker)
+    ///   - `Receiver<Packet<TestBuffer>>` a single queue for processed packets (worker -> dispatcher)
+    #[deny(clippy::type_complexity)]
+    #[deny(clippy::while_let_loop)]
+    fn spawn_workers(
+        workers: usize,
+        setup_pipeline: &Arc<dyn Send + Sync + Fn() -> DynPipeline<TestBuffer>>,
+    ) -> io::Result<(
+        Vec<chan::Sender<Packet<TestBuffer>>>,
+        chan::Receiver<Packet<TestBuffer>>,
+    )> {
+        let w = workers.max(1);
+        let (tx_from_workers, rx_from_workers) = chan::bounded::<Packet<TestBuffer>>(4096);
+        let mut to_workers = Vec::with_capacity(w);
+
+        for wid in 0..w {
+            let (tx_to_worker, rx_to_worker) = chan::bounded::<Packet<TestBuffer>>(8192);
+            to_workers.push(tx_to_worker);
+
+            let tx_out = tx_from_workers.clone();
+            let setup = setup_pipeline.clone();
+
+            let builder = std::thread::Builder::new().name(format!("dp-worker-{wid}"));
+            let handle_res = builder.spawn(move || {
+                let mut pipeline = (setup)();
+                loop {
+                    match rx_to_worker.recv() {
+                        Ok(pkt) => {
```

Comment:
Can you create an issue to batch packets for the kernel driver?  No need to fix it in this PR, but going packet by packet kind of sucks.  We should be able to send a collection of packets over instead, though this has its own performance considerations.

Reply:
I will create a separate issue